src/Microsoft.Extensions.DependencyInjection/ServiceProvider.cs

```diff
@@ -362,7 +368,15 @@ public override object Invoke(ServiceProvider provider)
 
             public override Expression Build(Expression provider)
             {
-                return base.Build(Expression.Field(provider, "_root"));
+                // PERF: Singleton services once created will always be available in the lookup dictionary
+                // Hence it is cheaper to directly call its Invoke method that will do the lookup rather than 
+                // building an expression tree that does the same.
+                // This change reduces the memory allocations during the start up because, expressions accessing private members
+                // involves expensive code access security checks for Desktop CLR.
+                return Expression.Call(
+                    Expression.Constant(this, typeof(SingletonCallSite)),
+                    InvokeMethodInfo,
+                    provider);
```

Comment:
To @pranavkm's point about singletons being added to the list of disposables, how does this work in those cases? Is it because invoke gets called before this ever does?